sonar/modules/documents/extensions.py

```diff
@@ -51,10 +53,35 @@ def pre_create(self, record):
         # Generate URN codes for documents without URNs.
         if not record.get_rero_urn_code(record):
             Urn.create_urn(record)
+        if record.model:
+            with db.session.begin_nested():
+                record.model.data = record
+                db.session.add(record.model)
 
     def post_create(self, record):
         """Called after a record is created.
 
         :param record: the invenio record instance to be processed.
         """
-        register_urn_code_from_document.delay(record)
+        try:
+            Urn.register_urn_code_from_document(record)
+        except Exception as e:
+            current_app.logger.error(f'Error during URN registration{e}')
+
+
+    def post_delete(self, record, force=False):
+        """Called after a record is deleted.
+
+        :param record: the invenio record instance to be processed.
+        :param force: unused.
+        """
+        from .api import DocumentRecord
+        if urn_code := DocumentRecord.get_rero_urn_code(record):
+            try:
+                urn_pid = PersistentIdentifier.get('urn', urn_code)
+                urn_pid.delete()
+                current_app.logger.error(
+                    f'Document (pid:{record["pid"]}) has been deleted and '
+                    f'contains an URN ({urn_code})')
+            except PIDDoesNotExistError:
+                pass
```

Comment:
Should we log a info message in this case?

Reply:
No, this is just for the record deletion, which never appens.